src/agent/rustjail/src/cgroups/fs/mod.rs

```diff
@@ -27,1120 +43,848 @@ macro_rules! sl {
-pub const MB: u128 = 1000 * KB;
-pub const GB: u128 = 1000 * MB;
-pub const TB: u128 = 1000 * GB;
-pub const PB: u128 = 1000 * TB;
-
-pub const KiB: u128 = 1024;
-pub const MiB: u128 = 1024 * KiB;
-pub const GiB: u128 = 1024 * MiB;
-pub const TiB: u128 = 1024 * GiB;
-pub const PiB: u128 = 1024 * TiB;
-
-pub const HUGETLB_BASE: &'static str = "hugetlb";
-pub const HUGETLB_USAGE: &'static str = "usage_in_bytes";
-pub const HUGETLB_MAX_USAGE: &'static str = "max_usage_in_bytes";
-pub const HUGETLB_FAILCNT: &'static str = "failcnt";
-
-fn parse_size(s: &str, m: &HashMap<String, u128>) -> Result<u128> {
-    let re = Regex::new(r"(?P<num>\d+)(?P<mul>[kKmMgGtTpP]?)[bB]?$")?;
-    let caps = re.captures(s).unwrap();
-
-    let num = caps.name("num");
-    let size: u128 = if num.is_some() {
-        num.unwrap().as_str().trim().parse::<u128>()?
-    } else {
-        return Err(nix::Error::Sys(Errno::EINVAL).into());
-    };
-
-    let q = caps.name("mul");
-    let mul: u128 = if q.is_some() {
-        let t = m.get(q.unwrap().as_str());
-        if t.is_some() {
-            *t.unwrap()
-        } else {
-            return Err(nix::Error::Sys(Errno::EINVAL).into());
-        }
+pub fn load<'a>(h: Box<&'a dyn cgroups::Hierarchy>, path: &str) -> Option<Cgroup<'a>> {
+    let valid_path = path.trim_start_matches("/").to_string();
+    let cg = cgroups::Cgroup::load(h, valid_path.as_str());
+    let cpu_controller: &CpuController = cg.controller_of().unwrap();
+    if cpu_controller.exists() {
+        Some(cg)
     } else {
-        return Err(nix::Error::Sys(Errno::EINVAL).into());
-    };
-
-    Ok(size * mul)
-}
-
-fn custom_size(mut size: f64, base: f64, m: &Vec<String>) -> String {
-    let mut i = 0;
-    while size >= base && i < m.len() - 1 {
-        size /= base;
-        i += 1;
+        None
     }
-
-    format!("{}{}", size, m[i].as_str())
 }
 
-pub const HUGEPAGESIZE_DIR: &'static str = "/sys/kernel/mm/hugepages";
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Manager {
+    pub paths: HashMap<String, String>,
+    pub mounts: HashMap<String, String>,
+    // pub rels: HashMap<String, String>,
+    pub cpath: String,
+}
 
-fn get_hugepage_sizes() -> Result<Vec<String>> {
-    let mut m = Vec::new();
-    for e in fs::read_dir(HUGEPAGESIZE_DIR)? {
-        let name = e?.file_name().into_string().unwrap();
-        let parts: Vec<&str> = name.split('-').collect();
-        if parts.len() != 2 {
-            continue;
-        }
-        let size = parse_size(parts[1], &BMAP)?;
-        m.push(custom_size(size as f64, 1024.0, DABBRS.as_ref()));
+impl CgroupManager for Manager {
+    fn apply(&self, pid: pid_t) -> Result<()> {
+        let h = cgroups::hierarchies::auto();
+        let h = Box::new(&*h);
+        let cg = load_or_create(h, &self.cpath);
+        cg.add_task(CgroupPid::from(pid as u64));
+        Ok(())
     }
 
-    Ok(m)
-}
+    fn set(&self, r: &LinuxResources, update: bool) -> Result<()> {
```

Comment:
This function is very long. Consider splitting it into per-controller sub-functions?